OADB/AliPhysicsSelection.cxx

```diff
@@ -251,63 +253,31 @@ UInt_t AliPhysicsSelection::CheckTriggerClass(const AliVEvent* event, const char
-  
-  while (1) {
-    AliDebug(AliLog::kDebug, trigger.Data());
-    
-    TArrayI pos;
-    Int_t nMatches = fRegexp->Match(trigger, "", 0, 2, &pos);
-    
-    if (nMatches <= 0) break;
-    
-    TString token(trigger(pos[0], pos[1]-pos[0]+1));
-
-    TParameter<Int_t>* param = dynamic_cast<TParameter<Int_t> *>(fCashedTokens->FindObject(token));
-    if (!param) {
-      TInterpreter::EErrorCode error;
-      Int_t bit = gInterpreter->ProcessLine(Form("AliTriggerAnalysis::k%s;", token.Data()), &error);
-      
-      if (error > 0) AliFatal(Form("Trigger token %s unknown", token.Data()));
-      
-      param = new TParameter<Int_t>(token, bit);
-      fCashedTokens->Add(param);
-      AliDebug(AliLog::kDebug, "Added token");
-    }
-    
-    Long64_t bit = param->GetVal();
-    
-    AliDebug(AliLog::kDebug, Form("Tok %d %d %s %lld", pos[0], pos[1], token.Data(), bit));
-    
-    if (offline) 
-      bit |= AliTriggerAnalysis::kOfflineFlag;
-    
-    if(event && triggerAnalysis) {
-      trigger.ReplaceAll(token, Form("%d", triggerAnalysis->EvaluateTrigger(event, (AliTriggerAnalysis::Trigger) bit)));
-    }
+/// Evaluate if the given event fulfills a given trigger logic
+///
+/// \param event Pointer to the current event
+/// \param triggerAnalysis Pointer to the TriggerAnlysis class
+/// \param triggerLogic Describing trigger logic; e.g. "V0A && V0C && ZDCTime && !TPCHVdip"
+/// \param offline Offline analysis(?)
+///
+/// \return True if the given event matches the trigger logic
+Bool_t AliPhysicsSelection::EvaluateTriggerLogic(const AliVEvent* event,
+						 AliTriggerAnalysis* triggerAnalysis,
+						 const char* triggerLogic, Bool_t offline){
+  auto formula_and_bits = FindForumla(triggerLogic);
+  auto trg_formula = formula_and_bits.first;
+  auto bits = formula_and_bits.second;
+  // Get the values for each individual trigger in the trigger logic string;
+  // These values are the parameters of the TFormula
+  std::vector<Double_t> paras(bits.size());
+  auto offline_flag = offline ? AliTriggerAnalysis::kOfflineFlag : 0;
+  for (size_t i = 0; i < bits.size(); ++i) {
+    typedef AliTriggerAnalysis::Trigger Trigger;
+    Trigger bit = static_cast<Trigger>(bits[i] | offline_flag);
+    paras[i] = triggerAnalysis->EvaluateTrigger(event, bit);
   }
-  
-#if ROOT_VERSION_CODE >= ROOT_VERSION(6,3,0)
-  // In case of ROOT6 it is necessary to stay for the moment with the v5 version of TFormula
-  // as the v6 version produces a large amount of warnings at runtime.
-  ROOT::v5::TFormula formula("formula",trigger);
-#else
-  TFormula formula("formula", trigger);
-#endif
-  if (formula.Compile() > 0)
-    AliFatal(Form("Could not evaluate trigger logic %s (evaluated to %s)", triggerLogic, trigger.Data()));
-  Bool_t result = formula.Eval(0);
-  
-  AliDebug(AliLog::kDebug, Form("%s --> %d", trigger.Data(), result));
-  
-  return result;
+  trg_formula.SetParameters(paras.data());
+  return trg_formula.Eval(0);
```

Comment:
`TFormula::SetParameters` copies the parameters into the `TFormula`'s member `fParams`. I don't think this copying is necessary, because they are used only once and will be reset when the `EvaluateTriggerLogic` is called next time.
In addition to `TFormula::Eval`, there is a `TFormula::EvalPar` method: https://root.cern.ch/doc/master/classROOT_1_1v5_1_1TFormula.html#a1f08c006a60991cc790d5c9410605787
It is defined as
```
virtual Double_t EvalPar(const Double_t *x, const Double_t *params=0){return ((*this).*fOptimal)(x,params);};
```
It uses pointer-to-member call syntax, but I think it is not difficult to guess that it points to one of the `TFormula`'s protected methods with suitable signature:
```
Double_t        EvalParFast(const Double_t *x, const Double_t *params);
Double_t        EvalPrimitive(const Double_t *x, const Double_t *params);
Double_t        EvalPrimitive0(const Double_t *x, const Double_t *params);
Double_t        EvalPrimitive1(const Double_t *x, const Double_t *params);
Double_t        EvalPrimitive2(const Double_t *x, const Double_t *params);
Double_t        EvalPrimitive3(const Double_t *x, const Double_t *params);
Double_t        EvalPrimitive4(const Double_t *x, const Double_t *params);
```
For example, the definition of the `EvalParFast` starts with the following lines
```
Double_t TFormula::EvalParFast(const Double_t *x, const Double_t *uparams)
 {
     const Double_t  *pdata[3] = {x,(uparams!=0)?uparams:fParams, fConst};
```
It uses the argument as an array of parameters if it is not null pointer.

So the calls to `SetParameters` and `Eval` can be merged into call to the `EvalPar`. It will not only remove unnecessary copying, but also take one line instead of two. It is not a premature optimization, it is removal of a premature pessimization.